rfcs/0013-configurable-and-composable-storage.md

```diff
@@ -0,0 +1,600 @@
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+The only drawback I can think of is the time and effort needed to implement the proposal and to deal with the [breaking changes](#breaking-changes). However, the impact of _not_ improving the current storage handling will over time surely be higher. It would mean carrying on the issues mentioned in the [Motivation](#motivation) and, worst of all, living with an API that is error-prone by-design.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+During the work on this RFC, five different approaches were intensely discussed and modeled in code. Non of them came even near to the simplicity of use and implementation of the proposed approach. Also, all of the discussed approaches could not guarantee an API that would prevent error-prone implementation of storage types, when it comes to the containment and composition semantics.
+
+For the storage API design guidelines, three different approaches were considered. We've tried to, e.g.:
+- additionally distinguish between operations failing because of the storage being uninitialized and because of semantic errors like out of bound access. 
+- avoid having possible multiple versions of a method by, e.g., putting the `Storage` instance in a "state" where it behaves differently for particular operations.
+
+All three approaches ended up to be less intuitive to use and implement then the approach proposed in the [Storage API design guidelines](#storage-api-design-guidelines).
+
+# Prior art
+
+[prior-art]: #prior-art
+
+As the [`storage_keys` RFC](./0008-storage-handler.md) puts it, there is no much prior art that would correspond with the overall direction for storage handling that we took in Sway. This means, having the `storage` elements be distinguished and having the API that communicates that storage operations might fail.
+
+The proposed approach builds on top of those two premisses, and brings robust and easy to implement composable storage types that can be arbitrarily configured in `storage` declarations.
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+Through the RFC process I expect the following open questions to be resolved:
+- questions posed in the [sample code](../files/0013-configurable-and-composable-storage/) and marked with `TODO-DISCUSSION`.
+- naming of proposed code elements like, e.g., method/function names, type names, etc. Since those names will become the part of the storage API it is of highest importance to come up with good and expressive names for abstractions.
+
+Also, I expect the current storage attributes, `#[storage(read, write)]`, to be discussed. In the sample code, the existing attributes are used. However, there are two questions to be decided on:
+- if the meaning of `write` remains "read and write" as it is now, the attributes should be renamed to `readonly` and `readwrite`.
```

Comment:
There are important reasons (for standards, notably) that we cannot be explicitly bound to read and write and must have read only/read write semantics.

Maybe we can just have the smallest change:
```sway
#[storage(read)]
#[storage(read_write)]
```